keyboards/bastardkb/charybdis/charybdis.c

```diff
@@ -0,0 +1,245 @@
+                if ((get_mods() | get_oneshot_mods()) & MOD_MASK_SHIFT) {
+                    keyboard_config.sniper_config = keyboard_config.sniper_config - 1;
+                } else {
+                    keyboard_config.sniper_config = keyboard_config.sniper_config + 1;
+                }
+                keyboard_config.sniper_config = constrain(keyboard_config.sniper_config, 0, SNIPER_OPTION_SIZE - 1);
+                eeconfig_update_kb(keyboard_config.raw);
+            }
+        break;
+
+        case DPI_CONFIG:
+            if(record->event.pressed){
+                if ((get_mods() | get_oneshot_mods()) & MOD_MASK_SHIFT) {
+                    keyboard_config.dpi_config = keyboard_config.dpi_config - 1;
+                } else {
+                    keyboard_config.dpi_config = keyboard_config.dpi_config + 1;
+                }
+                keyboard_config.dpi_config = constrain(keyboard_config.dpi_config, 0, DPI_OPTION_SIZE - 1);
+                eeconfig_update_kb(keyboard_config.raw);
+                pmw_set_cpi(dpi_array[keyboard_config.dpi_config]);
+            }
+        break;
+        
+        case DRAG_SCROLL:
+#ifdef CHARYBDIS_DRAGSCROLL_TOGGLE
+            if (record->event.pressed)
+#endif
+            {
+                is_drag_scroll ^= 1;
+            }
+#ifndef CHARYBDIS_DRAGSCROLL_VARIABLE
+            pmw_set_cpi(is_drag_scroll ? CHARYBDIS_DRAGSCROLL_DPI : dpi_array[keyboard_config.dpi_config]);
+#else
+            pmw_set_cpi(is_drag_scroll ? (dpi_array[keyboard_config.dpi_config] * CHARYBDIS_DRAGSCROLL_MULTIPLIER) : dpi_array[keyboard_config.dpi_config]);
+#endif
+        break;
+    }
+
+/* If Mousekeys is disabled, then use handle the mouse button
+ * keycodes.  This makes things simpler, and allows usage of
+ * the keycodes in a consistent manner.  But only do this if
+ * Mousekeys is not enable, so it's not handled twice.
+ */
+#ifndef MOUSEKEY_ENABLE
+    if (IS_MOUSEKEY_BUTTON(keycode)) {
+        report_mouse_t currentReport = pointing_device_get_report();
+        if (record->event.pressed) {
+            currentReport.buttons |= 1 << (keycode - KC_MS_BTN1);
+        } else {
+            currentReport.buttons &= ~(1 << (keycode - KC_MS_BTN1));
+        }
+        pointing_device_set_report(currentReport);
+        pointing_device_send();
+    }
+#endif
+
+    return true;
+}
+
+// Hardware Setup
+void keyboard_pre_init_kb(void) {
+    //debug_enable  = true;
+    //debug_matrix  = true;
+    //debug_mouse   = true;
+    scroll_inertia.x = 0;
+    scroll_inertia.y = 0;
+    // debug_encoder = true;
+    
+    keyboard_pre_init_user();
+}
+
+void pointing_device_init(void) {
+    pmw_spi_init();
+    pmw_set_cpi(dpi_array[keyboard_config.dpi_config]);
+}
+
+static bool has_report_changed(report_mouse_t new, report_mouse_t old) { return (new.buttons != old.buttons) || (new.x&& new.x != old.x) || (new.y&& new.y != old.y) || (new.h&& new.h != old.h) || (new.v&& new.v != old.v); }
```

Comment:
```suggestion
```

Reply:
`has_report_chaged` is called by `pointing_device_send` ?

Reply:
Yup!
https://github.com/qmk/qmk_firmware/blob/9c175f180138558748b0b240e76cb04fa11f9927/quantum/pointing_device.c#L34-L47

Reply:
yes, so deleting the function like you proposed doesn't make sense no ?

Reply:
The function is actually in pointing_device.c, as well.  So it's not needed here.

Reply:
allright, done. now calling `has_mouse_report_changed`